sc20 Revisions to scripting with spack (#93)
spack/spack-tutorial: tutorial_spack_scripting.rst

Scripting with Spack
====================

This tutorial introduces advanced capabilities for programmatically
querying information about packages installed with Spack so you can

> create your own custom scripts. Specifically, we will discuss how to

Reviewer: This first sentence is technically correct but hard to follow. Can we make it simpler, just say "this tutorial introduces advanced Spack features related to scripting"?

-----------------------
Setting up the tutorial
-----------------------

At this point you may have a lot of packages already installed,
depending on which sections of the tutorial you've done so far.
Let's ensure the outputs for this segment are reasonable using
commands presented in earlier sections of the tutorial.

Remove ``gcc@8.3.0`` and re-install ``hdf5`` and ``zlib@clang``
using the following commands:

.. literalinclude:: outputs/scripting/setup.out
   :language: console
   :emphasize-lines: 1,3,24

Now we are ready to use the Spack's ``find`` and ``python`` subcommands
to query the installed packages.

-----------------------------
Scripting with ``spack find``
-----------------------------

Earlier sections of the tutorial introduced different uses of the
``spack find`` command for querying installed packages at the command
line. Depending on the scripting language and associated tools you want
to use, you can take advantage of the following advanced options to
generate machine-readable output:

- ``--format FORMAT``; and
- ``--json``.

``FORMAT`` is a string that uses the same syntax expected by the Python

> ``format()`` command. The format string is passed to ``Spec.format``.

Reviewer: This isn't a command that the user should be expected to be familiar with at this point. At the very least, it should be `Spec.format()` to make clear it is not mirroring some builtin python capability.

entering the following two commands:

.. literalinclude:: outputs/scripting/spack-python-1.out
   :language: console
   :emphasize-lines: 1,4

Now let's revisit the Spack ``Spec`` object described in earlier tutorials.
We will also look at querying Spack's internal database of installed
packages. We will do our examples from within the python interpreter.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Accessing the ``Spec`` object
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Since the Spack API is available within the interpreter, we can access
both abstract and concrete specs. Typically ``package.py`` files reference
concrete specs within the ``install`` method. But we'll look at the
differences between the two.

Open the python interpreter with ``spack python``, instantiate the
``zlib`` spec, and check a few properties of an abstract spec:

.. literalinclude:: outputs/scripting/spack-python-abstract.out
   :language: console
   :emphasize-lines: 1-3,5,11,13

Notice that there are Spec properties and methods that are not accessible
to abstract specs; specifically:

- an exception -- ``SpecError`` -- is raised if we try to access its
  ``version``;
- there are no associated ``versions``; and

> - the platform operating system is ``None``.

Reviewer: ```suggestion
- the operating system is ``None``.
```

Now, from within the same interpreter session, let's concretize the spec
and try again:

.. literalinclude:: outputs/scripting/spack-python-concrete.out
   :language: console
   :emphasize-lines: 1-2,4,6,8

Notice that the concretized spec now:

- has a ``version``;
- has a single entry in its ``versions`` list; and
- the platform operating system is now ``ubuntu18.04``.

It is not necessary to store the intermediate abstract spec as we can
see when we instantiate and concretize a new instance:

.. literalinclude:: outputs/scripting/spack-python-sans-intermediate.out
   :language: console
   :emphasize-lines: 1-2

^^^^^^^^^^^^^^^^^^^^^^^^^^^
Querying the Spack database
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Even more powerful queries are available when we look at the information
stored in the Spack database, which manages read, write, and locking

> operations on the backing file system.

Reviewer: This is an implementation detail. From a user's perspective, it stores the full concrete spec of every package installed with Spack. I don't think how we accomplish that (locking, read/write locks, downgrades, etc) is relevant to the scripting.

> Most queries will be using the ``Database.query`` method on the singleton

Reviewer: Again, singleton is a detail that will distract from the point you're explaining at this point. How about:

"The spack database is accessible using ``spack.store.db``. Most queries will be using the ``Database.query`` method on the database."?

instance of the database, which is accessed using ``spack.store.db``.
Let's see the documentation available for the ``query`` method using 
python's built-in ``help``.

Enter the following python statements from within the interpreter:

.. literalinclude:: outputs/scripting/spack-python-db-query-help.out
   :language: console
   :emphasize-lines: 1-2,9-13

We will primarily make use of the ``query_spec`` argument.

Recall that queries using the ``spack find`` command are limited to 
queries of attributes with matching values, not values they do *not*
have. In other words, we cannot use the ``spack find`` command for
all packages that *do not* satisfy a certain criterion.

> Using the database, we can now perform such queries.

Reviewer: ```suggestion
Using the python interface, we can now perform such queries.
```
The database itself cannot be queried for negations either.

Author: Excellent point, thanks.

For example, let's find all packages that were compiled with ``gcc``

> but do not depend on ``mpich``. We can do this from within the 
> interpreter by entering the following python statements, where we
> deliberately use ``spack.cmd.display_specs`` to get the output in
> the same pretty-printed format as that of the ``spack find`` command:

Reviewer: This sentence feels like we're putting the technical details ahead of the point.

```suggestion
but do not depend on ``mpich``.  We can do this by using custom python
code and Spack database queries. We will use the ``spack.cmd.display_specs``
for output to achieve the same printing functionality as the ``spack find`` command.
```

.. literalinclude:: outputs/scripting/spack-python-db-query-exclude.out
   :language: console
   :emphasize-lines: 1-5

Now we have a powerful query not available through ``spack find``.

Now exit the interpreter so we are back at the command line:

.. code-block:: console

   >>> exit()

^^^^^^^^^^^^^
Using scripts
^^^^^^^^^^^^^

> Suppose we want to re-use our query with different specs?

Reviewer: I don't think we actually use any different specs in the tutorial. How about:

```suggestion
Suppose we want to re-use our query as medium-term functionality?
```

With a few generalizations to accept the include and exclude specs
on the command line, we can create a general-purpose query script.

Open a file called ``find_exclude.py`` in your preferred editor
and add the following code:

.. literalinclude:: outputs/scripting/0.find_exclude.py.example
   :language: python

Notice we added importing and using the system package (``sys``)
to access access the first and second command line arguments. We

> also concretized the exclude spec.

Reviewer: We don't concretize the exclude spec.

Author: Thanks for catching that.  I meant 'instantiated'.  (It was late for me.)

.. literalinclude:: outputs/scripting/find-exclude-1.out
   :language: console
   :emphasize-lines: 1

This is *great* for us, as long as we remember to use Spack's
``python`` command to run the script.

-------------------------------------
Using the ``spack-python`` executable
-------------------------------------

What if we want to make our script available for others to use without
the hassle of having to remember to use ``spack python``?

We can take advantage of the shebang line typically added as the
first line of python executable files. But there is a catch, as 
we will soon see.

Open the ``find_exclude.py`` script we created above in your preferred
editor and add the shebang line with ``spack python`` as the arguments
to ``env``:

.. literalinclude:: outputs/scripting/1.find_exclude.py.example
   :language: python
   :emphasize-lines: 1

Then exit our editor and add execute permissions to the script before
running it as follows:

.. literalinclude:: outputs/scripting/find-exclude-2.out
   :language: console
   :emphasize-lines: 1-2

If you are lucky, it worked on your system but there is no guarantee.
Some systems only support a single argument on the shebang line.

> So Spack developers created essentially a wrapper script for

Reviewer: ```suggestion
So Spack developers created spack-python, which isessentially a wrapper script for
```

Author: Thanks.  (Needs a space between 'is' and 'essentially'. ;)
